### PR TITLE
Move privileges resolution to Roles/RolesService

### PR DESCRIPTION
**Pull request by matriv**
- Use a `Map<String, Role>` instead of a list in `RolesService`. Keep the `List<Role>` in the `Roles` interface for easier testing but implement use a `Map<String, Role>` in `RolesService` internally and override the `findUser()` to do a fast map lookup instead of iterating over the list.
- Move privileges resolution to Roles/RolesService, this is a preparation step to make recursive resolution based also on granted roles (hierarchy) easier.

Relates: #12109

### CODE REVIEW

File: server/src/main/java/io/crate/metadata/shard/ShardReferenceResolver.java [hunk 1]
```diff
@@ -53,7 +53,7 @@ public class ShardReferenceResolver implements ReferenceResolver<NestableInput<?
 
     private static final Logger LOGGER = LogManager.getLogger(ShardReferenceResolver.class);
     private static final StaticTableReferenceResolver<ShardRowContext> SHARD_REFERENCE_RESOLVER_DELEGATE =
-        new StaticTableReferenceResolver<>(SysShardsTableInfo.create().expressions());
+        new StaticTableReferenceResolver<>(SysShardsTableInfo.create(List::of).expressions());
```
Comment (matriv) [inline server/src/main/java/io/crate/metadata/shard/ShardReferenceResolver.java:56]: Heads up, I think that there is no side-effect that we pass a no-op `Roles` here, but please also double check, maybe I'm missing something.

Comment (seut) [reply server/src/main/java/io/crate/metadata/shard/ShardReferenceResolver.java:56]: I think you're right as the shards are already protected by the `SysShardsTableInfo.getRouting()`.
But we should add least add that note to highlight it. Passing an empty list here without any further note looks dangerous and confusing.

**Comment by matriv** [comment]
retest this please

File: server/src/test/java/io/crate/role/PrivilegesTest.java [hunk 2]
```diff
@@ -32,42 +34,49 @@
...
         //ensureUserHasPrivilege will not throw an exception if the schema is `pg_catalog`
-        Privileges.ensureUserHasPrivilege(Privilege.Clazz.SCHEMA, "pg_catalog", USER);
-        Privileges.ensureUserHasPrivilege(Privilege.Clazz.TABLE, "pg_catalog.pg_am", USER);
-        Privileges.ensureUserHasPrivilege(Privilege.Clazz.TABLE, "pg_catalog.pg_database", USER);
+        ensureUserHasPrivilege(Privilege.Clazz.SCHEMA, "pg_catalog");
+        ensureUserHasPrivilege(Privilege.Clazz.TABLE, "pg_catalog.pg_am");
+        ensureUserHasPrivilege(Privilege.Clazz.TABLE, "pg_catalog.pg_database");
+    }
+
+    private void ensureUserHasPrivilege(Privilege.Clazz clazz, String ident) {
```
Comment (seut) [inline server/src/test/java/io/crate/role/PrivilegesTest.java:75]:
```suggestion
    private static void ensureUserHasPrivilege(Privilege.Clazz clazz, String ident) {
```

File: server/src/test/java/io/crate/role/PrivilegesTest.java [hunk 3]
```diff
@@ -32,42 +34,49 @@
...
+        Privileges.ensureUserHasPrivilege(ROLES, USER, clazz, ident);
+    }
+
+    private void ensureUserHasPrivilege(Privilege.Type type, Privilege.Clazz clazz, String ident) {
```
Comment (seut) [inline server/src/test/java/io/crate/role/PrivilegesTest.java:79]:
```suggestion
    private static void ensureUserHasPrivilege(Privilege.Type type, Privilege.Clazz clazz, String ident) {
```

File: libs/shared/src/main/java/io/crate/common/FourFunction.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * Represents a function that accepts three arguments and produces a result.
+ *
+ * @param <A> the type of the first argument
+ * @param <B> the type of the second argument
+ * @param <C> the type of the third argument
+ * @param <D> the type of the fourth argument
+ * @param <R> the return type
+ */
+@FunctionalInterface
+public interface FourFunction<A, B, C, D, R> {
```
[the remark this answers is not in the document]
Comment (matriv) [reply libs/shared/src/main/java/io/crate/common/FourFunction.java:34]: I also thought to just add this to `HasPrivilegeFunction` (the abstract one) but then, maybe one needs it in the future, and creates another "hidden" one, but also no strong opinion, happy to remove it.

**Review by seut — APPROVED** [review]
👍 thx

**Pull request merged**